cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -448,24 +570,32 @@ public boolean apply(RunRecordMeta input) {
     return getNonCompleteRuns(programRunIds, TYPE_RUN_RECORD_STARTED, limit);
   }
 
-  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(Set<ProgramRunId> programRunIds, String recordType,
+  private Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(Set<ProgramRunId> programRunIds, int limit) {
+    return getNonCompleteRuns(programRunIds, TYPE_RUN_RECORD_COMPLETED, limit);
+  }
+
+  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(Set<ProgramRunId> runIds, String recordType,
                                                               int limit) {
+    Set<ProgramRunId> programRunIds = new HashSet<>(runIds);
     Set<MDSKey> keySet = new HashSet<>();
     for (ProgramRunId programRunId : programRunIds) {
       keySet.add(getProgramKeyBuilder(recordType, programRunId).build());
+      if (programRunId.getVersion().equals(ApplicationId.DEFAULT_VERSION)) {
+        keySet.add(getVersionLessProgramKeyBuilder(recordType, programRunId).build());
+      }
     }
-    Map<MDSKey, RunRecordMeta> returnMap = listKV(keySet, RunRecordMeta.class, limit);
-    return getProgramRunIdMap(returnMap);
-  }
-
-  private Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(Set<ProgramRunId> programRunIds, int limit) {
-    Set<MDSKey> keySet = new HashSet<>();
 
-    for (ProgramRunId programRunId : programRunIds) {
-      keySet.add(getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programRunId).build());
-    }
     Map<MDSKey, RunRecordMeta> returnMap = listKV(keySet, RunRecordMeta.class, limit);
-    return getProgramRunIdMap(returnMap);
+    Map<ProgramRunId, RunRecordMeta> programRunIdRunRecordMetaMap = getProgramRunIdMap(returnMap);
+    Set<ProgramRunId> candidateSet = Sets.newHashSet(programRunIdRunRecordMetaMap.keySet());
+    for (ProgramRunId programRunId : candidateSet) {
+      if (!programRunIds.contains(programRunId)) {
+        // If the program run id, doesn't match with the requested set, remove it. This might happen
+        // getProgramKeyBuilder doesn't include the runid of the program.
```

Comment:
Even if `getProgramKeyBuilder` does not include the run id, it would be better to send the condition `programRunIds.contains(programRunId)` as a predicate to `listKV`